### DISCUSSION TITLE
es6 vs common js, 모듈의 역사

### DISCUSSION THREAD

**Discussion by hwangstar156** [woowacourse-study/2022-woowahan-bansanghwe#82] [JavaScript]
모듈의 역사와 es6와 common js의 차이

## 모듈이란?

모듈이란 별도로 분리된 재사용 가능한 코드조각을 의미합니다.

## 자바스크립트에서 왜 모듈이 나오게 된걸까?

원래의 자바스크립트는 모듈화와는 거리가 멀었다. 하지만 프론트엔드의 영역이 커지고 코드가 방대해지기 시작했다. 그래서 별도의 모듈로 분할하기위한 메커니즘의 필요성이 증대해졌습니다.

## 자바스크립트에서의 모듈의 역사

초기에는 `<script>` 태그로 불러오는 방식을 이용했다. 하지만 여기에는 여러가지 문제점이 있습니다.

- 전역 변수 공간이 오염이 가능하다.
>`<script>` 태그는 실제로 각각의 모듈 스코프를 가지고 있는것이 아니라 전역공간을 공유하기 때문에 예상치 못한 에러가 나올수 있다.
- 의존성이 제대로 관리하기가 힘들다.
> 모듈 a가 모듈 b에 의존한다면, 모듈 b가 로드되고 나서 모듈 a가 로드되어야 한다. 이러한 모듈관계가 수십, 수백개가 된다면 생산성이 떨어진다.
- 로드 시간
> 스크립트 태그는 새로운 HTTP 요청을 보낸다. HTTP2가 등장하기전에는 HTTP 요청이 병렬이 아니라 직렬로 요청되었다. 그래서 여러개의 script태그가 있다면 하나마다 HTTP요청을 한번씩 보내야 했다. 그래서 모든 스크립트가 다운 받아질떄까지 유저는 기다려야되는 문제가 있었다.

## IIFE

흔히 우리가 즉시 실행함수라고 부르는 IIFE는 자바스크립트에서 유사하게 모듈처럼 사용하기 위해서 나왔던 기법입니다.
```jsx
(function a() {}) ()
```
이런식으로 함수를 선언과 동시에 실행을 합니다. 이렇게 되면 함수 내부에 있는것들은 함수 스코프에 존재하기 때문에 전역공간을 사용하지 않을수 있습니다. 하지만 이렇게 이용을 하면 외부에서 즉시실행함수의 내부를 참조할수 없다는 문제가 있습니다.

이것을 해결하기 위해서 CommonJs가 해결책을 제시했습니다.

## CommonJs

CommonJs는 브라우저 뿐만아니라 서버사이드에서도 이용할수 있도록 만들어진 모듈화 표준입니다.

CommonJs는 다음과 같이 문제를 해결하였습니다.

- 모듈 파일마다 스코프가 설정되어있어서 변수 충돌이 일어나지 않는다.
- 파일 덩어리를 불러오는게 아니라 필요한 함수, 변수를 불러올수 있게 되었다.
- module.exports, require를 통해 의존되어있는 파일, 패키지들을 관리할수 있게 되었다.

하지만 CommonJs도 다음과 같은 문제점이 존재했습니다.

- 서버사이드에서는 js파일 마다 독립적인 스코프를 만들기때문에 전역공간과 분리가 되지만 브라우저에서 동작할떄는 파일마다 단일한 스코프가 존재하지 않기에 `<script>` 태그로 js파일을 로드하면 여전히 전역변수가 오염된다.
- CommonJs는 blocking방식이여서 브라우저에서는 사용자 경험에 문제 생길수도 있다.
```jsx
const foo = require('foo.js');
const bar = require('bar.js');
```

이렇게 모듈을 불러온다면 foo를 다 불러올때까지 bar는 불러오지 않게된다. 그래서 foo에서 불러오는 시간이 오래걸릴경우 전체적으로 느리게 모듈을 불러오는것 처럼 느껴지게 됩니다.

## AMD(asynchronous module definition)

비동기 상황에서도 자바스크립트 모듈을 쓰기 위해 CommonJs에서 논의하다가 합의점을 찾지못하고 분리한 그룹입니다.
브라우저에서의 non-blocking, async 성질을 적극 활용하여 웹에 적합한 모듈화 시스템을 만드려고 했습니다.

AMD를 구현한 모듈시스템으로는 `Require.js`가 있습니다.

- CommonJs와 동일하게 require, module.exports를 이용하기에 의존성 문제를 동일하게 해결할수 있다.
- 웹에서의 자바스크립트를 위해 나온 모듈이다.

AMD에는 어떤 문제점이 있어서 현재 이용하지 않는걸까??

- 문법이 복잡하다.

## UMD

CommonJs와 Amd는 서로 지향하는것이 달랐습니다. 이처럼 통일되지 않는 규격들의 발생으로 서로간의 호환성 문제로 이어졌습니다.

이떄문에 등장한것이 UMD입니다.  CommonJs와 AMD모두 호환할수 있도록 구현하였습니다.

## esmodule

UMD 역시 CommonJs와 Amd의 호환성만 해결할뿐 자바스크립트의 모듈시스템의 부재를 해결하지 못하였습니다. 

그래서 es6사양에서 드디어 javascript의 표준 모듈 시스템인 esmodule이 등장합니다.

esmodule의 특징은 다음과 같습니다.

- UMD, AMD와 다르게 문법이 간단합니다.
- 동기/비동기 로드를 모두 지원합니다.
- 정적 분석이 가능하기때문에 트리쉐이킹도 쉽게 가능해졌습니다.

이렇게 모든 문제점을 해결한 esmodule에도 문제점이 존재할까요?

네, 바로 esmodule은 위에서 말했듯 es6에서부터 나온 문법이기 때문에 구형브라우저에서는 안된다는 문제가 있습니다. 그렇다면 다시 새로운 모듈시스템이 나왔을까요??

## 모듈 번들러

이것을 해결하기 위해서 모듈번들러라는것이 나왔습니다. 도구들로는 webpack, rollup, parcel등등이 있습니다.

**Comment by airman5573** [comment]
모듈은 왜 필요한가요?

**Comment by airman5573** [comment]
자바스크립트의 모듈에 대해서 설명해주세요

**Comment by airman5573** [comment]
모듈과 함수의 차이는 무엇인가요?

**Comment by airman5573** [comment]
전역 공간이 오염되는 이유만으로 모듈이 나온건가요?

**Comment by airman5573** [comment]
전역 변수가 오염되는게 왜 안좋은건가요?

**Comment by airman5573** [comment]
전역 변수를 오염한다는게 구체적으로 어떤건가요?

**Comment by airman5573** [comment]
window객체를 오염한다는건 어떤 의미인가요?

**Comment by airman5573** [comment]
non-blocking/blocking에 대해서 설명해 주세요

**Comment by airman5573** [comment]
defer vs async

**Comment by airman5573** [comment]
defer와 async가 나오게된 배경은 무엇인가요?

**Comment by airman5573** [comment]
왜 defer/async를 기본동작으로 하지 않았을까요?

**Comment by airman5573** [comment]
왜 스크립트 태그를 만나면 html파싱을 멈출까요?

**Comment by airman5573** [comment]
브라우저의 렌더링 과정에 대하여 간단하게 설명해 주세요

**Comment by airman5573** [comment]
composite 과정에 대해 설명해 주세요

**Comment by airman5573** [comment]
파싱의 정의가 뭘까요?

**Comment by airman5573** [comment]
브라우저는 import / export 코드를 어떻게 처리하나요?

**Comment by airman5573** [comment]
es6 module을 사용하신 이유가 있나요?

**Comment by airman5573** [comment]
AMD/UMD가 있음에도 es6 module이 나온 이유가 무엇인가요?
